### PR TITLE
First version of an SQW writer

### PR DESCRIPTION
**Pull request by jl-wynen**
This is not very user friendly because it requires a lot of interdependent inputs. See also the many TODO comments.

~Opening this now to get the CI running. It's not quite ready for review.~

Ready for review, I think. But it will need iterating on in the future.

### CODE REVIEW

File: src/scippneutron/io/sqw/_build.py [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+        self,
+    ) -> tuple[
+        dict[DataBlockName, memoryview | None],
+        dict[DataBlockName, SqwDataBlockDescriptor],
+    ]:
+        data_blocks = self._prepare_data_blocks()
+        buffers: dict[DataBlockName, memoryview | None] = {}
+        descriptors = {}
+        for name, data_block in data_blocks.items():
+            buffer = BytesIO()
```
Comment (SimonHeybrock) [inline src/scippneutron/io/sqw/_build.py:214]: Can you add some comment explaining why we have a `BytesIO` here? Or maybe it belongs into the class docstring?

File: src/scippneutron/io/sqw/_models.py [hunk 2]
```diff
@@ -0,0 +1,457 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
```
Comment (SimonHeybrock) [inline src/scippneutron/io/sqw/_models.py:2]: Here and elsewhere: Some module-level (and lower) documentation about the overall pieces and methodology would be useful. As it is now I can see that all the bits and bobs are nicely written, but it is not possible to tell how it is working overall.

File: src/scippneutron/io/sqw/_models.py [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+
+    def _serialize_to_dict(
+        self,
+    ) -> dict[str, ir.Object | ir.ObjectArray | ir.CellArray]:
+        return {
+            "serial_name": ir.String(self.serial_name),
+            "version": ir.F64(self.version),
+            "name": ir.String(self.name),
+            "target_name": ir.String(self.target_name),
+            "frequency": ir.F64(self.frequency.value),  # TODO unit
```
Comment (SimonHeybrock) [inline src/scippneutron/io/sqw/_models.py:257]: todo?

File: src/scippneutron/io/sqw/_models.py [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+
+
+@dataclass(kw_only=True, slots=True)
+class SqwIXSource(ir.Serializable):
+    name: str
+    target_name: str
+    frequency: sc.Variable
+
+    serial_name: ClassVar[str] = "IX_source"
+    version: ClassVar[float] = 2.0
```
Comment (SimonHeybrock) [inline src/scippneutron/io/sqw/_models.py:247]: Do different pieces in the file have different versions?

File: src/scippneutron/io/sqw/_models.py [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+class EnergyMode(enum.Enum):
+    direct = 1
+    indirect = 2
+
+
+@dataclass(kw_only=True, slots=True)
+class SqwIXSample(ir.Serializable):
+    name: str
+    lattice_spacing: sc.Variable  # vector
+    lattice_angle: sc.Variable  # vector
```
Comment (SimonHeybrock) [inline src/scippneutron/io/sqw/_models.py:289]: I presume this means `vector3`?

File: src/scippneutron/io/sqw/_models.py [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+                shape=(len(self.indices),),
+                # +1 to convert to 1-based indexing
+                data=np.array(self.indices) + 1.0,
+                ty=ir.TypeTag.f64,
+            ),
+        }
+
+
+def _angle_value(x: sc.Variable) -> float:
+    return x.to(unit="rad", dtype="float64", copy=False).value  # type: ignore[no-any-return]
```
Comment (SimonHeybrock) [inline src/scippneutron/io/sqw/_models.py:429]:
```suggestion
    return float(x.to(unit="rad", dtype="float64", copy=False))
```
might avoid the mypy bypass? Or does that raise if not dimensionless? So maybe divide by `sc.scalar(1, unit='rad')`? Or leave it, not too important.

File: src/scippneutron/io/sqw/_sqw.py [hunk 7]
```diff
@@ -0,0 +1,562 @@
...
+
+    @property
+    def byteorder(self) -> Byteorder:
+        return self._sqw_io.byteorder
+
+    def data_block_names(self) -> Iterable[DataBlockName]:
+        """Return an iterator over the names of all stored data blocks."""
+        return self._block_allocation_table.keys()
+
+    # TODO lock blocks during writing, esp pix data
```
Comment (SimonHeybrock) [inline src/scippneutron/io/sqw/_sqw.py:101]: todo?

**Review by SimonHeybrock — COMMENTED** [review]
Impressive!

Code looks clean, but it really needs to lower-level and higher-level docstrings for anyone to be able to see through the complexities.

File: src/scippneutron/io/sqw/_ir.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+
+Object = Struct | String | F64 | U64 | U32 | U8 | Logical | Array | Datetime
+
+
+class Serializable(ABC):
+    @abstractmethod
+    def _serialize_to_dict(self) -> dict[str, Object | ObjectArray | CellArray]: ...
+
+    def serialize_to_ir(self) -> Struct:
```
[the remark this answers is not in the document]
Comment (jl-wynen) [reply src/scippneutron/io/sqw/_ir.py:125]: 'Intermediate Representation'.

File: src/scippneutron/io/sqw/_read_write.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+
+_READERS = _IORegistry[_ObjectReader]("reader")
+_WRITERS = _IORegistry[_ObjectWriter]("writer")
+
+
+def read_object_array(sqw_io: LowLevelSqw) -> ir.ObjectArray | ir.CellArray:
+    position = sqw_io.position
+    ty = ir.TypeTag(sqw_io.read_u8())
+    if ty == ir.TypeTag.serializable:  # TODO
+        # raise RuntimeError(f'!!!! {ty.value} {sqw_io.position-1}')
```
[the remark this answers is not in the document]
Comment (jl-wynen) [reply src/scippneutron/io/sqw/_read_write.py:55]: This `raise` is a left over that should be removed. But the larger issue is refactoring how self-serializing types work. It is a messy hack at the moment.

File: src/scippneutron/io/sqw/_models.py (hunk 5 above)
Comment (jl-wynen) [reply src/scippneutron/io/sqw/_models.py:289]: Yes

File: src/scippneutron/io/sqw/_build.py [hunk 10]
```diff
@@ -0,0 +1,413 @@
...
+        self._dnd_placeholder: _DndPlaceholder | None = None
+        self._pix_placeholder: _PixPlaceholder | None = None
+        self._instrument: SqwIXNullInstrument | None = None
+        self._sample: SqwIXSample | None = None
+
+    @contextmanager
+    def create(self) -> Generator[Sqw, None, None]:
+        from ._sqw import Sqw
+
+        # TODO better mechanism that doesn't use "r+"
```
[the remark this answers is not in the document]
Comment (jl-wynen) [reply src/scippneutron/io/sqw/_build.py:94]: I would like to first see how the current interface works in a practical example. And then, I'd like some input on how best to implement this.
My idea was to get the current mechanism merged adn then iterate on it. But if you prefer, we could discuss the implementation first.

[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply src/scippneutron/io/sqw/_build.py:94]: I am fine with iterating later, just asking!

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**